File: typescript/matrix/matrix.ts
export class Matrix {
  _rows: Array<number[]>
  _columns: Array<number[]>

  constructor(raw: string) {
    this._rows = this.rowsFromRaw(raw)
    this._columns = this.transpose(this._rows)
  }

  get rows(): Array<number[]> {
    return this._rows
  }

  get columns(): Array<number[]> {
    return this._columns
  }

  rowsFromRaw(raw: string): Array<number[]> {
    const rawRows = raw.split('\n')
    return rawRows.reduce((rows: Array<number[]>, current) => {
      const row = current.split(' ').map(n => parseInt(n))
      return [...rows, row]
    }, [])
  }

  transpose(matrix: Array<number[]>): Array<number[]> {
    return matrix[0].map((_, i) => {
      return matrix.map((row => row[i]))
    })
  }
}

File: ruby/matrix/matrix.rb
class Matrix
  def initialize(raw)
    @rows = parse(raw)

    # hand rolled for practice, but could just use @rows.transpose
    @columns = transpose(@rows)
  end

  def row(num)
    @rows[num - 1]
  end

  def column(num)
    @columns[num - 1]
  end

  private

  def parse(raw)
    raw.split("\n").map do |row_str|
      row_str.split(' ').map(&:to_i)
    end
  end

  def transpose(matrix)
    num_cols = matrix[0].length

    (0...num_cols).map do |index|
      matrix.map do |row|
        row[index]
      end
    end
  end
end
